### DISCUSSION TITLE
What conditions do custom types need to meet in order to use [1.. ^ 1]

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/csharplang#8279] [Q&A] [answered]
I found that both string and Memory can use this syntax. But I couldn't find it in the API documentation
```
string this[Range] { get; }
```

**Comment by KennethHoff** [comment]
It's the syntax for the `System.Range` type.

**Comment by 333fred** [comment]
See https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges#implicit-range-support.

**Comment by ParsaMehdipour** [comment]
@zms9110750 
The code you mentioned is an indexer that uses the Range type. This feature was introduced in C# 8.0 as part of the Index and Range feature set. It allows for more convenient slicing of strings and collections.
You're correct that this syntax isn't explicitly documented in the API for string or Memory<T>. This is because it's a language feature rather than a specific method on these types. The C# compiler translates this syntax into appropriate method calls behind the scenes.
For strings, this indexer is translated into calls to the Substring method. For Memory<T>, it's translated into calls to the Slice method.

**Reply by zms9110750** [reply]
Does this mean that if I customize a type, I can only define it through an indexer?

**Reply by ParsaMehdipour** [reply, marked as answer]
You can check the link to the documentation itself: 
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges#implicit-range-support
